### PR TITLE
Add Submit Ballot API

### LINKED ISSUE #135: Create Ballot Uploader Service
Create a ballot uploader service to upload `SubmittedBallot` from electionguard and store in data storage using a message bus. 

CiphertextBallot -> Upload Ballot -> Success or Failure message
SubmittedBallot -> Cast/Spoil Ballot -> Success or Failure message

Relies on #137 and #136

### PR DESCRIPTION
**Pull request by AddressXception**
refactor cache and queue into interfaces.  Refactor base models.  Fix linting configuration and issues, separate files, update postman tests

### Issue

Fixes #135

### Description
Add submit ballot api's and refactor for a consistent api surface across the related ballot api's.  This PR lays the groundwork for future PR's by factoring different concerns into separate files and adding a paradigm for handling errors as well as validating data.  it touches a lot of file but all of the work is in the ballot api segment.

notice specifically that the ballot api's make use of `post` requests for cast/spoil operations whereas it uses a `put` for submit since submit is supposed to be idempotent.  another design consideration is that submit accepts the election_id as either a query parameter or in the request body.  either must be provided but the query string will override the body.

also, the handling of data context state such as elections and contexts is not yet implemented from the cache since those api's do not exist.  For this reason, our api's also support providing the election description and the context in the body of the request.  If they are provided in the body then the method is stateless (currently) but if they are not provided, they would be looked up from the data cache (in the future).

lastly, these api's should be atomic in the sense that all validation should be done prior to committing any state changes

regarding the repository/cache and the queue, default implementations are provided for mongo and rabbit, however memory implementations are also provided.  This is largely just for testing purposes.  No global application instances are created in memory so a new object is returned for each request (meaning you can't submit a ballot and then get it in a different request when running the application on raw hardware). We can add that in the future if it is useful.

### Testing
1. `make docker-dev`
2. using postman run: `submit ballot` and `get ballot`
3. run `validate ballot` and `validate manifest`

Expected: all of the sample data validates correctly and the submitted ballot in the sample data is fetched from mongo in the get call

### CODE REVIEW

File: app/api/v1/mediator/decrypt.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from fastapi import APIRouter, Body, HTTPException
+
+from ..models import DecryptBallotsWithSharesRequest
+from ..tags import ENCRYPT
+
+router = APIRouter()
+
+
+@router.post("/decrypt", tags=[ENCRYPT])
+def decrypt_ballots(request: DecryptBallotsWithSharesRequest = Body(...)) -> Any:
```
Comment (AddressXception) [inline app/api/v1/mediator/decrypt.py:19]: this is just a move refactor

File: app/api/v1/mediator/encrypt.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    EncryptBallotsRequest,
+    EncryptBallotsResponse,
+)
+from ..tags import ENCRYPT
+
+router = APIRouter()
+
+
+@router.post("/encrypt", tags=[ENCRYPT])
+def encrypt_ballots(
```
Comment (AddressXception) [inline app/api/v1/mediator/encrypt.py:23]: also just a move refactor

File: app/api/v1/mediator/manifest.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+from app.core.schema import get_description_schema
+
+from ..models import ValidateManifestRequest, BaseValidationResponse, ResponseStatus
+from ..tags import MANIFEST
+
+router = APIRouter()
+
+
+@router.post("/validate", tags=[MANIFEST], status_code=status.HTTP_204_NO_CONTENT)
+def validate_election_manifest(
```
Comment (AddressXception) [inline app/api/v1/mediator/manifest.py:17]: also just a move refactor

File: app/api/v1/mediator/ballot.py [hunk 4]
```diff
@@ -1,221 +1,262 @@
...
-    Construct a lookup by ballot ID containing the dictionary of shares needed
-    to decrypt that ballot.
-    """
-    shares_by_ballot: Dict[str, Dict[str, DecryptionShare]] = {}
-    for share in shares:
-        ballot_shares = shares_by_ballot.setdefault(share.object_id, {})
-        ballot_shares[share.guardian_id] = share
+def _get_election_parameters(
+    election_id: Optional[str], request: BaseBallotRequest
+) -> Tuple[Manifest, CiphertextElectionContext, str]:
```
Comment (AddressXception) [inline app/api/v1/mediator/ballot.py:151]: in a future PR this function will try to query for the Manifest and Context from the mongo repository if they are not provided in the request body

File: app/core/queue.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    "IMessageQueue",
+    "MemoryMessageQueue",
+    "RabbitMQMessageQueue",
+    "get_message_queue",
+]
+
+settings = Settings()
+
+
+class IMessageQueue(Protocol):
```
Comment (AddressXception) [inline app/core/queue.py:19]: the queue interface is by no means complete, it's just something simple to get started.

File: app/core/repository.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+from .settings import Settings, StorageMode
+
+__all__ = ["IRepository", "MemoryRepository", "MongoRepository", "get_repository"]
+
+settings = Settings()
+
+DOCUMEMNT_VALUE_TYPE = Union[MutableMapping, List[MutableMapping]]
+
+
+class IRepository(Protocol):
```
Comment (AddressXception) [inline app/core/repository.py:16]: similarly this interface is also not complete, just something to get started

File: pyproject.toml [hunk 7]
```diff
@@ -30,8 +30,30 @@ requests = "^2.24.0"
 target-version = ['py38']
 include = '\.pyi?$'
 
+[tool.pylint.'FORMAT']
+max-line-length = 120
+
 [tool.pylint.'MESSAGES CONTROL']
-disable = "duplicate-code, missing-module-docstring, missing-class-docstring, missing-function-docstring, logging-fstring-interpolation"
+disable = '''
+    duplicate-code,
```
Comment (AddressXception) [inline pyproject.toml:38]: match the python repo

File: app/core/repository.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+from pymongo import MongoClient
+from pymongo.database import Database
+
+from .settings import Settings, StorageMode
+
+__all__ = ["IRepository", "MemoryRepository", "MongoRepository", "get_repository"]
+
+settings = Settings()
+
+DOCUMEMNT_VALUE_TYPE = Union[MutableMapping, List[MutableMapping]]
```
Comment (SteveMaier-IRT) [inline app/core/repository.py:13]: misspelled?
DOCUMEMNT_VALUE_TYPE

File: app/api/v1/mediator/ballot.py [hunk 9]
```diff
@@ -1,221 +1,262 @@
...
-            routing_key="submitted-ballots",
-            body=json.dumps(ballot.to_json_object()),
-        )
-        channel.close()
-        connection.close()
-    except (pika.exceptions.ChannelError, pika.exceptions.StreamLostError):
+        with get_repository(election_id, DataCollection.SUBMITTED_BALLOT) as repository:
+            cacheable_ballots = [ballot.to_json_object() for ballot in ballots]
+            keys = repository.set(cacheable_ballots)
+            print(keys)
```
Comment (SteveMaier-IRT) [inline app/api/v1/mediator/ballot.py:188]: Do we really want to print keys here or was this just for debugging?

**Review by SteveMaier-IRT — CHANGES_REQUESTED** [review]

File: app/api/v1/mediator/ballot.py (hunk 9 above)
Comment (AddressXception) [reply app/api/v1/mediator/ballot.py:188]: it was for debugging, removing

**Review by SteveMaier-IRT — APPROVED** [review]

**Pull request merged**